### ISSUE TITLE
JSON Patch in GameServer Controller is Lossy on Int64

### ISSUE THREAD

**Issue by igooch** [googleforgames/agones#3636]
**TL;DR**:
JSON does not support Int64 well. The current recommendation is that anyone using the `count` or `capacity` values should avoid using extremely large numbers.

**Issue with setting count or capacity for Counter to max(int64)**: 

When the Counter `capacity` or `count` are set to max(int64) in the game server spec and running `kubectl create -f fleet.yaml` we get an error: 
`spec.template.spec.counters.sessions.capacity: Invalid value: "number": spec.template.spec.counters.sessions.capacity in body must be of type integer: "number"`
 
When the Counter `capacity` or `count` in the game server spec are close to, but not at, max(int64) `kubectl create` does not error. However, when checking the details of the game server with `kubectl describe` the values are truncated:
```
spec:
  template:
    spec:
      counters:
        sessions:
          count: 8611686018427387904
          capacity: 8611686018427387904
```
becomes
```
Spec:
  Template:
    Spec:
      Counters:
        Sessions:
          Capacity:  8611686018427388000
          Count:     8611686018427388000
```

**Debugging**:

@zmerlynn root-caused the issue #3608, and found that the source issue is with `jsonpatch` translating values to scientific notation. A different repo ran into the same error and has a longer description of the issue https://github.com/evanphx/json-patch/issues/189. We're unable to switch to this repo because the fixed method `CreateMergePatch` just gives back bytes, but in tests, etc. we use the jsonpatch structures. This will likely need to be an upstream fix in [gomodules/jsonpatch](https://github.com/gomodules/jsonpatch) or [kubernetes-sigs/json](https://github.com/kubernetes-sigs/json). The takeaway is that JSON does not support int64 well, and if the controller is having this many issues there are likely other issues that would crop up elsewhere in the code, and other languages would have yet another set of issues.

**Comment by zmerlynn** [comment]
Filed https://github.com/gomodules/jsonpatch/issues/39

**Comment by markmandel** [comment]
Keeping this open to track the upstream work.

**Comment by lacroixthomas** [comment]
Hello,
I was looking into this issue, seems that the https://github.com/gomodules/jsonpatch is a fork of https://github.com/mattbaird/jsonpatch which is working fine with the int64 max values, the fork is currently outdated (13 commits behind)
Is there a specific reason agones is using the fork instead of the main repos ? (maybe for the go modules ?)

**Comment by markmandel** [comment]
> https://github.com/mattbaird/jsonpatch which is working fine with the int64 max values

From memory, it was not working with int64 max values at the time we tested it. Looking at the commit history, I'm not seeing anything that indicates a fix - but please correct me if we're wrong.

> the fork is currently outdated (13 commits behind)

...and 39 commits ahead. So they diverged in implementation.

It also made sense to align with controller-runtime. See implementing patch #3639 for more details.

**Comment by lacroixthomas** [comment]
My bad, I wasn't testing it properly 😄 
Did some more test: https://go.dev/play/p/UanCz7xwaiB 
Using the `json.NewDecoder()` with `UseNumber()` (https://pkg.go.dev/encoding/json#Decoder.UseNumber) seems to be enough

I will check the history of this issue to understand a bit more, will try to open a PR on the https://github.com/gomodules/jsonpatch

**Comment by markmandel** [comment]
> I will check the history of this issue to understand a bit more, will try to open a PR on the https://github.com/gomodules/jsonpatch

Awesome. That'll be a good win for everyone!

**Comment by lacroixthomas** [comment]
Opened an MR on the jsonpatch repo: https://github.com/gomodules/jsonpatch/pull/40
But just realised that the repo doesn't seems to be maintained anymore, last changes were a couple of years ago, not sure if it's going to be reviewed, will do some tests on agones with my fork though, shall see
